tests/test-sessions/test-sessions-common/src/test/java/org/eclipse/jetty/server/session/AbstractSessionCacheTest.java

```diff
@@ -65,7 +126,56 @@ public void sessionDidActivate(HttpSessionEvent se)
     public abstract AbstractSessionCacheFactory newSessionCacheFactory(int evictionPolicy, boolean saveOnCreate, 
                                                                        boolean saveOnInactiveEvict, boolean removeUnloadableSessions,
                                                                        boolean flushOnResponseCommit);
-    
+
+    /**
+     * Test that a session that exists in the datastore, but that cannot be
+     * read will be invalidated and deleted, and thus a request to re-use that
+     * same id will not succeed.
+     * 
+     * @throws Exception
+     */
+    @Test
+    public void testUnreadableSession() throws Exception
+    {
+        Server server = new Server();
+        server.setSessionIdManager(new DefaultSessionIdManager(server));
+
+        ServletContextHandler context = new ServletContextHandler(ServletContextHandler.SESSIONS);
+        context.setContextPath("/test");
+        context.setServer(server);
+        server.setHandler(context);
+
+        AbstractSessionCacheFactory cacheFactory = newSessionCacheFactory(SessionCache.NEVER_EVICT, false, false, false, false);
+        SessionCache cache = cacheFactory.getSessionCache(context.getSessionHandler());
+
+        //prefill the datastore with a session that will be treated as unreadable
+        UnreadableSessionDataStore store = new UnreadableSessionDataStore(1, new SessionData("1234", "/test", "0.0.0.0", System.currentTimeMillis(), 0,0, -1));
+        cache.setSessionDataStore(store);
+        context.getSessionHandler().setSessionCache(cache);
+        server.start();
+
+        try (StacklessLogging stackless = new StacklessLogging(Log.getLogger("org.eclipse.jetty.server.session")))
+        {
+            //check that session 1234 cannot be read, ie returns null AND
+            //that it is deleted in the datastore
+            Session session = context.getSessionHandler().getSession("1234");
+            assertNull(session);
+            assertTrue(store._deletedIds.contains("1234"));
+
+            //now try to make a session with the same id as if from a request with
+            //a SESSION_ID cookie set
+            Request request = new Request(null, null);
+            request.setRequestedSessionId("1234");
+            HttpSession newSession = context.getSessionHandler().newHttpSession(request);
+            assertFalse("1234".equals(newSession.getId()));
```

Comment:
Replace with `assertNotEquals()`.

Reply:
Please add a comment here about the fact that since the session has been deleted, its id cannot be reused, and delete the javadoc of the method (which basically says the same, but in the wrong spot - and who reads test javadocs anyway?!? :smile:)